File: src/lib/store.ts
import fs from "node:fs";
import { LISTINGS_PATH, atomicWrite, quarantineCorrupt } from "./paths";
import type { Listing } from "./types";

// Simple file-backed store. Single-process, fine for a self-hosted client tool.
function readAll(): Listing[] {
  try {
    if (!fs.existsSync(LISTINGS_PATH)) return [];
    return JSON.parse(fs.readFileSync(LISTINGS_PATH, "utf8"));
  } catch {
    // parse error on an existing file → move it aside, don't overwrite history
    quarantineCorrupt(LISTINGS_PATH);
    return [];
  }
}

function writeAll(listings: Listing[]) {
  atomicWrite(LISTINGS_PATH, JSON.stringify(listings, null, 2));
}

export function listListings(): Listing[] {
  return readAll().sort((a, b) => b.createdAt.localeCompare(a.createdAt));
}

export function getListing(id: string): Listing | undefined {
  return readAll().find((l) => l.id === id);
}

export function saveListing(listing: Listing): Listing {
  const all = readAll();
  const idx = all.findIndex((l) => l.id === listing.id);
  listing.updatedAt = new Date().toISOString();
  if (idx >= 0) all[idx] = listing;
  else all.push(listing);
  writeAll(all);
  return listing;
}

export function updateListing(
  id: string,
  patch: Partial<Listing>
): Listing | undefined {
  const all = readAll();
  const idx = all.findIndex((l) => l.id === id);
  if (idx < 0) return undefined;
  all[idx] = { ...all[idx], ...patch, updatedAt: new Date().toISOString() };
  writeAll(all);
  return all[idx];
}

export function deleteListing(id: string): boolean {
  const all = readAll();
  const next = all.filter((l) => l.id !== id);
  if (next.length === all.length) return false;
  writeAll(next);
  return true;
}

File: src-tauri/src/store.rs
#![allow(dead_code)]
//! File-backed listings store (ported from src/lib/store.ts).

use crate::paths::{atomic_write, quarantine_corrupt, Paths};
use crate::types::Listing;
use crate::util::now_iso;
use std::sync::{Mutex, OnceLock};

/// Process-wide guard held across the read+write of listings.json, so concurrent
/// mutators (batch worker + user commands on the multi-threaded runtime) can't
/// last-writer-wins each other (which could revert a just-saved nmID → orphan).
fn store_lock() -> &'static Mutex<()> {
    static L: OnceLock<Mutex<()>> = OnceLock::new();
    L.get_or_init(|| Mutex::new(()))
}

fn read_all(paths: &Paths) -> Vec<Listing> {
    let file = paths.listings();
    let txt = match std::fs::read_to_string(&file) {
        Ok(t) => t,
        Err(_) => return vec![],
    };
    match serde_json::from_str::<Vec<Listing>>(&txt) {
        Ok(v) => v,
        Err(_) => {
            quarantine_corrupt(&file);
            vec![]
        }
    }
}

fn write_all(paths: &Paths, list: &[Listing]) {
    let _ = atomic_write(&paths.listings(), &serde_json::to_vec_pretty(list).unwrap_or_default());
}

pub fn list_listings(paths: &Paths) -> Vec<Listing> {
    let mut v = read_all(paths);
    v.sort_by(|a, b| b.created_at.cmp(&a.created_at));
    v
}

pub fn get_listing(paths: &Paths, id: &str) -> Option<Listing> {
    read_all(paths).into_iter().find(|l| l.id == id)
}

pub fn save_listing(paths: &Paths, mut listing: Listing) -> Listing {
    let _g = store_lock().lock().unwrap_or_else(|e| e.into_inner());
    let mut all = read_all(paths);
    listing.updated_at = now_iso();
    if let Some(i) = all.iter().position(|l| l.id == listing.id) {
        all[i] = listing.clone();
    } else {
        all.push(listing.clone());
    }
    write_all(paths, &all);
    listing
}

pub fn update_listing<F: FnOnce(&mut Listing)>(paths: &Paths, id: &str, f: F) -> Option<Listing> {
    let _g = store_lock().lock().unwrap_or_else(|e| e.into_inner());
    let mut all = read_all(paths);
    let i = all.iter().position(|l| l.id == id)?;
    f(&mut all[i]);
    all[i].updated_at = now_iso();
    let updated = all[i].clone();
    write_all(paths, &all);
    Some(updated)
}

pub fn delete_listing(paths: &Paths, id: &str) -> bool {
    let _g = store_lock().lock().unwrap_or_else(|e| e.into_inner());
    let mut all = read_all(paths);
    let n = all.len();
    all.retain(|l| l.id != id);
    if all.len() == n {
        return false;
    }
    write_all(paths, &all);
    true
}
